### PR TITLE
Implement the Core Entities

### PR DESCRIPTION
**Pull request by iamrecursion**
### Pull Request Description
This PR adds the raw entities that make up the Enso Core language. They are tested just for their basic properties and aren't yet really _used_.

### Important Notes
This PR intentionally _does not_ add benchmarks or higher-level tests. It makes far more sense to add these to the more complex Core constructs that will be implemented on top of this code, and as such the next task will do so. 

The PR seems very large, but ~1000 of the added lines are in the test file and are essentially rote. 

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md), [Java](https://github.com/luna/enso/blob/master/doc/java-style-guide.md), [Rust](https://github.com/luna/enso/blob/master/doc/rust-style-guide.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/haskell-style-guide.md) style guides as appropriate.
- [x] All code has been tested where possible.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -384,22 +413,26 @@ lazy val runtime = (project in file("engine/runtime"))
...
+      "org.graalvm.truffle" % "truffle-dsl-processor" % graalVersion % "provided",
+      "org.graalvm.truffle" % "truffle-tck"           % graalVersion % "provided",
+      "org.graalvm.truffle" % "truffle-tck-common"    % graalVersion % "provided",
+      "org.scalacheck"      %% "scalacheck"           % "1.14.0" % Test,
+      "org.scalactic"       %% "scalactic"            % "3.0.8" % Test,
+      "org.scalatest"       %% "scalatest"            % "3.2.0-SNAP10" % Test,
+      "org.graalvm.truffle" % "truffle-api"           % graalVersion % Benchmark,
+      "org.typelevel"       %% "cats-core"            % "2.0.0-M4"
+    ),
+    Compile / unmanagedClasspath += (core_definition / Compile / packageBin).value,
```
Comment (kustosz) [inline build.sbt]: Please document this hack.

File: common/graph/src/main/scala/org/enso/graph/Graph.scala [hunk 2]
```diff
@@ -486,23 +556,84 @@ object Graph {
...
+      * No bounds checking is performed on any of the provided indices.
+      *
+      * @param componentIx the index representing the component [[C]]
+      * @param fieldIx the index of the field [[F]] in the component [[C]]
+      * @param ev evidence that component [[C]] has field [[F]] in [[G]]
+      * @tparam C the component type
+      * @tparam F the field type
+      * @return the value of field [[F]] in [[component]]
+      */
+    def primUnsafeReadField[C <: Component, F <: Component.Field](
```
Comment (kustosz) [inline common/graph/src/main/scala/org/enso/graph/Graph.scala]: The `prim` naming is unclear to me. Can this be named such that the difference between `unsafeReadField` and `primUnsafeReadField` is understandable from the name?

Comment (kustosz) [reply common/graph/src/main/scala/org/enso/graph/Graph.scala]: The same applies to `primWrite` all the whole family.

File: common/graph/src/main/scala/org/enso/graph/definition/Macro.scala [hunk 3]
```diff
@@ -442,6 +442,22 @@ object Macro {
...
       }
 
+      /** Makes a type accessor name from a type name.
+        *
+        * This basically means converting from `PascalCase` to `camelCase`.
+        *
+        * @param fieldName the name of the field
+        * @return the accessor name for [[fieldName]]
+        */
+      def makeTypeAccessorName(fieldName: TypeName): String = {
```
Comment (kustosz) [inline common/graph/src/main/scala/org/enso/graph/definition/Macro.scala]: Given we already use apache-commons in `runtime`, you could use `StringUtils.decapitalize`

File: common/graph/src/main/scala/org/enso/graph/definition/Macro.scala [hunk 4]
```diff
@@ -1276,6 +1296,9 @@ object Macro {
     *   val backref: mutable.Map[Int, Vector[Int]] = mutable.Map()
     * }
     * }}}
+    *
+    * Please note that while you can define a field with multiple opaque
```
Comment (kustosz) [inline common/graph/src/main/scala/org/enso/graph/definition/Macro.scala]: I think it's better to stick to impersonal voice in docs. I may be wrong, so leaving that up to you.

File: engine/core-definition/src/main/scala/org/enso/core/CoreGraph.scala [hunk 5]
```diff
@@ -0,0 +1,623 @@
...
+          *
+          * @param text a link to the [[RawLiteral]] representing the number
+          */
+        case class TextLiteral(text: Link[G])
+
+        /** The raw representation of a name.
+          *
+          * @param literal the literal text of the name
+          */
+        case class NameLiteral(literal: OpaqueData[String, NameStorage])
```
Comment (kustosz) [inline engine/core-definition/src/main/scala/org/enso/core/CoreGraph.scala]: What is it? I.e. what code generates this?

File: engine/core-definition/src/main/scala/org/enso/core/CoreGraph.scala [hunk 6]
```diff
@@ -0,0 +1,623 @@
...
+        /** A node representing boolean false. */
+        case class False()
+
+        // === Literals =======================================================
+
+        /** A raw literal is the basic literal type in the [[CoreGraph]].
+          *
+          * @param literal the literal text
+          */
+        case class RawLiteral(literal: OpaqueData[String, LiteralStorage])
```
Comment (kustosz) [inline engine/core-definition/src/main/scala/org/enso/core/CoreGraph.scala]: What is it? Is it a superclass of all literals or a raw-text-literal? If the former, please explain, if the latter please rename to `RawTextLiteral`.

File: engine/core-definition/src/main/scala/org/enso/core/CoreGraph.scala [hunk 7]
```diff
@@ -0,0 +1,623 @@
...
+        case class AtomDef(name: Link[G], args: Link[G])
+
+        /** An expanded-form type definition, with a body.
+          *
+          * @param name the name of the aggregate type
+          * @param typeParams the type parameters to the definition
+          * @param body the body of the type definition, represented as a
+          *             [[List]] of bindings
+          */
+        case class ComplexTypeDef(
```
Comment (kustosz) [inline engine/core-definition/src/main/scala/org/enso/core/CoreGraph.scala]: Why not just `TypeDef`?

File: engine/runtime/src/test/scala/org/enso/compiler/test/core/CorePrimTest.scala [hunk 8]
```diff
@@ -0,0 +1,958 @@
...
+      case ForeignDefinition.any(n1) =>
+        n1.language = l1
+        n1.code = l2
+
+        n1.foreignDefinition shouldEqual ForeignDefinitionVal(l1, l2)
+      case _ => fail
+    }
+  }
+
+  nodeShape should "be able to represent syntax errors" in {
```
Comment (kustosz) [inline engine/runtime/src/test/scala/org/enso/compiler/test/core/CorePrimTest.scala]: Those tests... Are they all necessary? I dislike this structure, it's a 1000 lines of self-evident stuff (like, you're essentially testing the macro system here), while it lacks any more involved logic, like constructing a few levels of AST, traversing it, having a non-tree-like structure and traversing that. All these tests prove is "there's a bunch of constructors that can be used", but do not actually tell anything about the deeper properties of Core.

File: engine/runtime/src/test/scala/org/enso/interpreter/test/instrument/ValueExtractorTest.scala [hunk 9]
```diff
@@ -11,6 +11,7 @@ class ValueExtractorTest extends InterpreterTest {
     val results = mutable.HashMap[(Int, Int), Any]()
     getValueExtractorInstrument.bindTo("Test.main", 7, 5, { res =>
       results.put((7, 5), res)
+      ()
```
Comment (kustosz) [inline engine/runtime/src/test/scala/org/enso/interpreter/test/instrument/ValueExtractorTest.scala]: Whyyyyyyyy?

File: engine/core-definition/src/main/scala/org/enso/core/CoreGraph.scala [hunk 10]
```diff
@@ -0,0 +1,619 @@
...
+        case class TopLevelBinding(module: Link[G], binding: Link[G])
+
+        // === Type Definitions ===============================================
+
+        /** An atom definition.
+          *
+          * @param name the name of the atom
+          * @param args the atom's arguments as a [[List]]
+          */
+        case class AtomDef(name: Link[G], args: Link[G])
```
Comment (wdanilo) [inline engine/core-definition/src/main/scala/org/enso/core/CoreGraph.scala]: I find the names not consistent. There is `AtomDef` but not `ModuleDef`. I would use `Atom` instead.

File: engine/core-definition/src/main/scala/org/enso/core/CoreGraph.scala [hunk 11]
```diff
@@ -0,0 +1,619 @@
...
+        )
+
+        // === Typing =========================================================
+
+        /** A type signature.
+          *
+          * @param typed the expression being ascribed a type
+          * @param sig the signature being ascribed to [[typed]]
+          */
+        case class TypeAscription(typed: Link[G], sig: Link[G])
```
Comment (wdanilo) [inline engine/core-definition/src/main/scala/org/enso/core/CoreGraph.scala]: Why its not named `TypeSignature`? I believe it is much more common word for that, especially as "ascription" means something different than signature - ascription "ascribes" signature to a variable. Oh, I just answered myself, the name is good :D

File: engine/core-definition/src/main/scala/org/enso/core/CoreGraph.scala [hunk 12]
```diff
@@ -0,0 +1,619 @@
...
+          * @param right the right type in the subsumption judgement
+          */
+        case class TypesetSubsumption(left: Link[G], right: Link[G])
+
+        /** The typeset equality judgement `~`.
+          *
+          * @param left the left operand
+          * @param right the right operand
+          */
+        case class TypesetEquality(left: Link[G], right: Link[G])
```
Comment (wdanilo) [inline engine/core-definition/src/main/scala/org/enso/core/CoreGraph.scala]: We do not have it in user-facing syntax, right?

File: engine/core-definition/src/main/scala/org/enso/core/CoreGraph.scala [hunk 13]
```diff
@@ -0,0 +1,619 @@
...
+          * @param right the right operand
+          */
+        case class TypesetIntersection(left: Link[G], right: Link[G])
+
+        /** The typeset subtraction operator `\`.
+          *
+          * @param left the left operand
+          * @param right the right operand
+          */
+        case class TypesetSubtraction(left: Link[G], right: Link[G])
```
Comment (wdanilo) [inline engine/core-definition/src/main/scala/org/enso/core/CoreGraph.scala]: I dont like keeping them in such form here. These should be just normal operators. `&` should parse to operator and should not be a special thing there. You should just see that `&` is defined on every type.

File: engine/core-definition/src/main/scala/org/enso/core/CoreGraph.scala [hunk 14]
```diff
@@ -0,0 +1,619 @@
...
+          * @param argument the argument to the function
+          */
+        case class Application(function: Link[G], argument: Link[G])
+
+        /** A mixfix function application.
+          *
+          * @param function the name of the mixfix function
+          * @param arguments the arguments to the mixfix function as a [[List]]
+          */
+        case class MixfixApplication(function: Link[G], arguments: Link[G])
```
Comment (wdanilo) [inline engine/core-definition/src/main/scala/org/enso/core/CoreGraph.scala]: Im wondering if we need it. Each `MixfixApplication` like `if a then b else c` thranslates to just `Application` in form of `if_then_else a b c` . I believe that keeping only `Application` would simplify the design here a lot.

File: engine/core-definition/src/main/scala/org/enso/core/CoreGraph.scala [hunk 15]
```diff
@@ -0,0 +1,619 @@
...
+          */
+        case class RightSection(operator: Link[G], arg: Link[G])
+
+        /** A centre section operator application.
+          *
+          * @param operator the operator being sectioned
+          */
+        case class CentreSection(operator: Link[G])
+
+        /** A representatin of a term that is explicitly forced.
```
Comment (wdanilo) [inline engine/core-definition/src/main/scala/org/enso/core/CoreGraph.scala]: Needs better description of what is "explicitly forced". Even if its temporary, Id love to understand it! :)

File: engine/core-definition/src/main/scala/org/enso/core/CoreGraph.scala [hunk 16]
```diff
@@ -0,0 +1,619 @@
...
+          * of lambdas.
+          */
+        case class LambdaShorthandArgument()
+
+        /** A function call-site argument.
+          *
+          * @param expression the argument expression
+          * @param name the name of the argument, if given
+          */
+        case class CallSiteArgument(expression: Link[G], name: Link[G])
```
Comment (wdanilo) [inline engine/core-definition/src/main/scala/org/enso/core/CoreGraph.scala]: What is this? Is it like calling `foo a b c` ? If so, what is the `name` when using unnamed args ?

File: engine/core-definition/src/main/scala/org/enso/core/CoreGraph.scala [hunk 17]
```diff
@@ -0,0 +1,619 @@
...
+          *                        of the scrutinee
+          */
+        case class TypeMatch(matchExpression: Link[G])
+
+        /** A pattern that matches on the scrutinee based on a type subsumption
+          * judgement and assigns a new name to it for use in the branch.
+          *
+          * @param matchExpression the expression representing the possible type
+          *                        of the scrutinee, and its new name
+          */
```
Comment (wdanilo) [inline engine/core-definition/src/main/scala/org/enso/core/CoreGraph.scala]: How is it used? I mean, if I write:

```haskell
case foo of
    Vector Int -> ...
    Vector.V3 x y z -> ...
```

Then you do not know after parsing which branch is type match and which one is constructor match. Is this meant to be used after some desugaring / typechecking passes?

File: engine/core-definition/src/main/scala/org/enso/core/CoreGraph.scala [hunk 18]
```diff
@@ -0,0 +1,619 @@
...
+          * @param doc a [[TextLiteral]] containing the documentation comment
+          */
+        case class DocComment(commented: Link[G], doc: Link[G])
+
+        /** A disable comment.
+          *
+          * @param disabledExpr the portion of the program that has been
+          *                     disabled
+          */
+        case class DisableComment(disabledExpr: Link[G])
```
Comment (wdanilo) [inline engine/core-definition/src/main/scala/org/enso/core/CoreGraph.scala]: There are no disable comments. Disable comments are macros.

File: engine/core-definition/src/main/scala/org/enso/core/CoreGraph.scala [hunk 19]
```diff
@@ -0,0 +1,619 @@
...
+      /** Checks whether a given node represents syntactic sugar.
+        *
+        * @param node the node to check
+        * @return `true` if [[node]] represents syntax sugar, `false` otherwise
+        */
+      def shapeIsSugar(
+        node: Node[CoreGraph]
+      )(implicit graph: PrimGraph.GraphData[CoreGraph]): Boolean = {
+        node match {
+          case Shape.TypeDef.any(_)    => true
```
Comment (wdanilo) [inline engine/core-definition/src/main/scala/org/enso/core/CoreGraph.scala]: strange alignment

File: engine/core-definition/src/main/scala/org/enso/core/CoreGraph.scala [hunk 20]
```diff
@@ -0,0 +1,619 @@
...
+        *
+        * @param source the node at the start of the link
+        * @param target the node at the end of the link
+        * @tparam G the graph type
+        */
+      @field case class Shape[G <: PrimGraph](source: Node[G], target: Node[G])
+
+      // ======================================================================
+      // === Utility Functions ================================================
+      // ======================================================================
```
Comment (wdanilo) [inline engine/core-definition/src/main/scala/org/enso/core/CoreGraph.scala]: remove

File: engine/runtime/src/test/scala/org/enso/compiler/test/core/CoreTest.scala [hunk 21]
```diff
@@ -0,0 +1,15 @@
...
+import org.enso.compiler.test.CompilerTest
+
+// TODO [AA] Tests of the actual core usage in semi-realistic scenarios:
+//  - Diamond
+//  - Multi-level tree
+//  - Larger diamonds
+//  - Cycle
+//  - Linked List
+//  - Parent Link walk to same place
+//  - etc.
```
Comment (wdanilo) [inline engine/runtime/src/test/scala/org/enso/compiler/test/core/CoreTest.scala]: Just a note - do not do it now. Do it after you implement real usages and preferably first integration with Graal. I believe that some of these shapes will never happen in real usages

Comment (kustosz) [reply engine/runtime/src/test/scala/org/enso/compiler/test/core/CoreTest.scala]: Counter point: do it now (i.e. next PR), to the extent the graph is guaranteed to support, code should have its contract tested when it lands. I'm OK with constraining the tests to DAGs – I agree supporting cycles is not necessary, but paths, multi-level trees and arbitrary diamonds are sure to happen and should be tested as soon as the API allows.

**Pull request merged**